src/vis/layers-factory.js

```diff
@@ -95,17 +95,31 @@ var LAYER_CONSTRUCTORS = {
   },
 
   cartodb: function (attrs, options) {
-    checkProperties(attrs, ['source', 'cartocss']);
-    attrs.source = CartoDBLayer.getLayerSourceFromAttrs(attrs, options.vis.analysis);
+    // TODO: Once https://github.com/CartoDB/cartodb/issues/12885 is merged,
+    // we should make 'source' attribute required again and make sure it's
+    // always populated:
+    // checkProperties(attrs, ['source', 'cartocss']);
+    // attrs.source = CartoDBLayer.getLayerSourceFromAttrs(attrs, options.vis.analysis);
+    checkProperties(attrs, ['cartocss']);
+    if (attrs.source) {
+      attrs.source = CartoDBLayer.getLayerSourceFromAttrs(attrs, options.vis.analysis);
+    }
 
     return new CartoDBLayer(attrs, {
       vis: options.vis
     });
   },
 
   torque: function (attrs, options) {
-    checkProperties(attrs, ['source', 'cartocss']);
-    attrs.source = TorqueLayer.getLayerSourceFromAttrs(attrs, options.vis.analysis);
+    // TODO: Once https://github.com/CartoDB/cartodb/issues/12885 is merged,
+    // we should make 'source' attribute required again and make sure it's
+    // populated:
+    // checkProperties(attrs, ['source', 'cartocss']);
+    // attrs.source = CartoDBLayer.getLayerSourceFromAttrs(attrs, options.vis.analysis);
+    checkProperties(attrs, ['cartocss']);
+    if (attrs.source) {
+      attrs.source = CartoDBLayer.getLayerSourceFromAttrs(attrs, options.vis.analysis);
```

Comment:
We're using `TorqueLayer.getLayerSourceFromAttrs`...

`CartoDBLayer.getLayerSourceFromAttrs` and `TorqueLayer.getLayerSourceFromAttrs` are identical so perhaps we could unify them somewhere. Perhaps they could live in the `AnalysisService` and here we could just do something like `options.vis.analysis.getSourceOrFindById(attrs.source)`.

Since this epic will be merged after `sources-as-references`, perhaps this is a good opportunity to get rid of the deprecation warnings in `getLayerSourceFromAttrs`.